runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,224 @@
+
+/**
+ * A simple thread based subscription engine where each subscriber or listener gets
+ * a thread which will listen on the tables of interest to it and moves at a rate
+ * which is the same as the consumption.
+ *
+ * Created by hisundar on 04/28/2020.
+ */
+@Slf4j
+public class StreamManager {
+
+    private static final int MAX_SUBSCRIBERS = 12; // Arbitrary sanity check value.
+
+    /**
+     * Corfu Runtime.
+     */
+    @Getter
+    private final CorfuRuntime runtime;
+
+    /**
+     * Map of StreamingSubscriptionThreads.
+     * <p>
+     * This is the map of all streaming subscriptions.
+     */
+    @Getter
+    private final Map<StreamListener, SubscriptionThread> subscriptions = new HashMap<>();
+
+    public StreamManager(@Nonnull CorfuRuntime runtime) {
+        this.runtime = runtime;
+    }
+
+    /**
+     * Subscribe to updates.
+     *
+     * @param streamListener   Client listener.
+     * @param namespace        Namespace of interest.
+     * @param tablesOfInterest Only updates from these tables will be returned.
+     * @param startAddress     Address to start the notifications from.
+     */
+    public synchronized <K extends Message, V extends Message, M extends Message>
+    void subscribe(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                   @Nonnull List<TableSchema<K, V, M>> tablesOfInterest, long startAddress) {
+        if (subscriptions.size() >= MAX_SUBSCRIBERS) {
+            throw new StreamSubscriptionException(
+                    "subscribe: Max " + MAX_SUBSCRIBERS + " hit! Consider consolidating "
+                            + streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new StreamSubscriptionException(
+                    "StreamManager:subscriber already registered "
+                            + streamListener + ". Maybe call unregister first?");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}",
+                streamListener, startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for {}", streamListener.toString());
+            return;
+        }
+        st.shutdown.set(true);
+        st.interrupt();
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+    }
+
+    class SubscriptionThread extends Thread {
```

Comment:
Spawning a thread per subscriber seems to be too expensive for what's being achieved. 
I think its better to multiplex these over an executor. The polling logic and life-cycle management is also cleaner with a scheduler executor. You can submit a periodic job that syncs the subscription streams, that will give you a task future that can be canceled. I think that's cleaner than managing the life cycle of the thread yourself. 

I think the code already has some bugs related to the thread life cycle management, so maybe consider just using a scheduled executor.

Reply:
this was the same comment by @xnull and while it is valid - the plan was to specifically limit this to one thread per subscriber for 
1. keeps the memory taken by a subscriber completely local to a thread
2. limits the number of threads/subscribers very explicitly

but the pool idea is definitely valid, I had planned to use it after we have stream tagging because that will change the threading model again.

Reply:
A scheduled executor will achieve point 1 and 2. I just don't see a strong reason for not using a scheduled executor. 

Might be missing something, but I don't see this as an intermediate step and I don't see it's relationship with stream tagging.